### ISSUE TITLE
Nullable success

### ISSUE THREAD

**Issue by lhengl** [Flutterando/result_dart#13]
The Success type S must extend an Object, which means that it is not possible to return a valid nullable success.

Why not allow nullable success? I understand that this is probably because we want to be explicit about avoiding the use of null values and treat null values as a failure? But what if a null value is actually a success?

For example, when you want to fetch a record from a database, if it doesn't exist, you'd want to return null instead of an exception. This is because we want to differentiate between a successful call that returns null, compared to an unsuccessful call such as a response code other than 200. At the moment they both have to be returned as failures.

Why does S and F even need to extend Object? How would you deal with a valid nullable value?

As a workaround I wrap the Success value with a NullableSuccess(T? value) where the value is a nullable value. But this is verbose for such a simple case.

The other workaround is to be explicit and implement a NotFound failure... But this means I have to remember to check on every call to handle this failure. It would be better to rely on Dart analyzer to tell me that it is nullable...

I think the best solution would be to remove the "S extends Object" in the package.

**Comment by lhengl** [comment]
I forked the repo and I tried to fix this myself, and it seems that the "swap" function prevents Success to be nullable because a Failure must extend Object.

What is a use case for swap? Why would there ever be a case where you want to swap a success with a failure and vice versa?

In the fork, I've removed the swap functions, which would be a breaking change if it is being used. But I honestly can't think of a reason why you would need that feature. So I've incremented the version to 2.0.0 to mark it as breaking change.

The fork is located here: [https://github.com/lhengl/result_dart](https://github.com/lhengl/result_dart)

Let me know if you agree, and I can create a PR.

Thank you.

**Comment by lhengl** [comment]
I decided that it would be a better idea to throw a NotFound exception instead as that is more explicit. Please ignore this issue. Thank you.

**Comment by micheljung** [comment]
> I decided that it would be a better idea to throw a NotFound exception instead as that is more explicit. Please ignore this issue. Thank you.

I disagree. Exceptions are expensive and shouldn't be used for normal flow control.

**Comment by micheljung** [comment]
@lhengl can you give an example of how this looks like? Mixing two concepts, "no value" and "something went wrong" seems bad to me and, AFAIK, makes code needlessly complicated.

**Comment by lhengl** [comment]
@micheljung sorry I haven't responded. I had to revisit this issue again. And you're right mixing errors and success of no value has been problematic, in all calls I had to check that there is a NotFoundError mixed in with all other errors. I learned this the hard way because now all my code is verbose with NotFoundError checks to ensure it is valid instead of throwing an error.

I will reopen this issue based on my experience and findings.

As a quick fix, I turned back to using a custom Nullable class to wrap the result value:

```
class NullableSuccess<T> {
  final T? value;
  const Nullable(this.value);
}

// use it like this
final result = Success(NullableSuccess(object));

// get the null value like this
final object = result.getOrThrow().value;
if (object == null) {
  // do something
}
```

**Comment by Bartozo** [comment]
Hi 👋, I was wondering why this isn’t possible with the latest version of this package. I am currently working on a feature where I need to fetch data for a user and check if they have an active collaboration in the system. In this case it is perfectly fine for the result to be empty. The absence of data is not an error in my scenario, so I would like to return success with an empty value.

This is exactly what I would like to see, but sadly this isn't possible with the current implementation of `ResultDart`, which extends from the `Object`
```
/// Gets an active collaboration for the client.
Future<ResultDart<Collaboration?, MyCustomException>> getActiveCollaborationForClient({required int clientId});
```

Is there a reason this pattern isn’t supported or recommended in result_dart? The only possible solution for now is probably the one mentioned by @lhengl, but I would like to avoid having an extra classes for that 😅

**Comment by lhengl** [comment]
Using Nullable has proven to be much of a pain as it was to handle it explicitly as error.

I've finally had enough of this issue and went ahead and forked it. It's now fixed. You can use nullable types. I have created a pull request. But if you want to help test it the fork is here: [https://github.com/lhengl/result_dart/](https://github.com/lhengl/result_dart/)

Side note: I've also updated the default result to use Object as a failure rather than Exception because this will allow you to do something like this:

```
try {
} catch (e,s) {
}

// otherwise it will be verbose like this:

try {
} on Exception catch (e,s) {
}
```

**Comment by jacobaraujo7** [comment]
I believe this would bring more problems than solutions, so it is not possible to implement this at the moment.
I’m closing this issue for now, but feel free to reopen it if needed.

**Comment by lhengl** [comment]
You mentioned: **This change intentionally makes it possible, but not encouraged by default, to use Success<S?>.** However, the package actually forbids it entirely — it’s not just discouraged, it’s enforced. If it is discouraged, then it would be allowed but documented as discouraged.

If the package is going to be opinionated in disallowing Success(null), then I believe it should also be explicit in how this case is meant to be handled. Right now, developers are left to guess the intended approach. While the API enforces the restriction, it doesn’t offer any built-in mechanism or clear documentation for representing a "success with no value" scenario — a very common case in practice.

That lack of guidance is what’s confusing. If the philosophy is to promote explicit modeling of states like “not found,” then the package should formalize that — either through a standard wrapper (like what you did with Unit for void), or at least through documentation and examples.

I understand the intention to enforce explicit handling of nullable success values. But the library currently pushes developers away from Success(null) without guiding them toward a consistent alternative.

In an opinionated package, this creates a gap: you’re enforcing a direction, but not offering a path. If null as a success is disallowed, there should be a clearly defined way — whether via a built-in wrapper like Nullable<T>, or documented idioms — to model common scenarios like "value not found but not an error."

Otherwise, each developer ends up inventing their own solution, which works against the clarity and consistency that opinionated libraries are supposed to promote.
